src/platform/linux_usbfs/usbfs.rs

```diff
@@ -39,6 +39,98 @@ pub fn release_interface<Fd: AsFd>(fd: Fd, interface: u8) -> io::Result<()> {
     }
 }
 
+#[repr(C)]
+struct DetachAndClaim {
+    interface: c_uint,
+    flags: c_uint,
+    driver: [c_uchar; 255 + 1],
+}
+
+pub fn detach_and_claim_interface<Fd: AsFd>(fd: Fd, interface: u8) -> io::Result<()> {
+    const USBDEVFS_DISCONNECT_CLAIM_EXCEPT_DRIVER: c_uint = 0x02;
+    unsafe {
+        let mut dc = DetachAndClaim {
+            interface: interface.into(),
+            flags: USBDEVFS_DISCONNECT_CLAIM_EXCEPT_DRIVER,
+            driver: [0; 256],
+        };
+
+        dc.driver[0..6].copy_from_slice(b"usbfs\0");
+
+        let ctl =
+            ioctl::Setter::<ioctl::ReadOpcode<b'U', 27, DetachAndClaim>, DetachAndClaim>::new(dc);
+        match ioctl::ioctl(&fd, ctl) {
+            Err(e) if e == io::Errno::NOTTY => {}
+            other => return other,
+        }
+
+        // disconnect-and-claim failed, fall back to the racy detach-and-claim
+        let detach_result = detach_kernel_driver(&fd, interface);
+
+        if let Err(e) = claim_interface(fd, interface) {
+            if let Err(detach_err) = detach_result {
+                return Err(detach_err);
+            }
+
+            return Err(e);
+        }
+
+        Ok(())
+    }
+}
+
+#[repr(C)]
+struct UsbFsIoctl {
+    interface: c_uint,
+    ioctl_code: c_uint,
+    data: *mut c_void,
+}
+
+#[repr(C)]
+struct GetDriver {
+    interface: c_int,
+    driver: [u8; 256],
+}
+
+pub fn detach_kernel_driver<Fd: AsFd>(fd: Fd, interface: u8) -> io::Result<()> {
+    unsafe {
+        // Only detach usbfs
+        let mut driver = GetDriver {
+            interface: interface.into(),
+            driver: [0; 256],
+        };
+        let ctl =
+            ioctl::Updater::<ioctl::ReadOpcode<b'U', 8, GetDriver>, GetDriver>::new(&mut driver);
+        ioctl::ioctl(&fd, ctl)?;
+
+        if driver.driver[0..6] != *b"usbfs\0" {
+            return Err(io::Errno::NOENT);
+        }
```

Comment:
This seems backwards -- you want to detach a kernel driver (e.g. `ftdi_sio`), but don't want to detach another instance of nusb / libusb using `usbfs`.

Reply:
Monkey see [this](https://github.com/libusb/libusb/blob/master/libusb/os/linux_usbfs.c#L1742), monkey do `!=`. Monkey no think source is C and that strcmp returns 0 if equal :)